tamusers/user_utils.py

```diff
@@ -49,55 +62,98 @@ def update_user(user, payload, oidc=False):
     if changed or not user.pk:
         user.save()
 
-    ad_groups = payload.get('ad_groups', None)
+    ad_groups = payload.get("ad_groups", None)
     # Only update AD groups if it's a list of non-empty strings
     if isinstance(ad_groups, list) and (
-            all([isinstance(x, str) and x for x in ad_groups])):
+        all([isinstance(x, str) and x for x in ad_groups])
+    ):
         user.update_ad_groups(ad_groups)
 
 
+@transaction.atomic()
 def get_or_create_user(payload, oidc=False):
-    user_id = payload.get('sub')
+    user_id = payload.get("sub")
     if not user_id:
-        msg = _('Invalid payload.')
+        msg = _("Invalid payload.")
         raise exceptions.AuthenticationFailed(msg)
 
     user_model = get_user_model()
 
-    with transaction.atomic():
-        try:
-            user = user_model.objects.select_for_update().get(uuid=user_id)
-        except user_model.DoesNotExist:
-            user = user_model(uuid=user_id)
-            user.set_unusable_password()
-        update_user(user, payload, oidc)
-
-    # If allauth.socialaccount is installed, create the SocialAcount
-    # that corresponds to this user. Otherwise logins through
-    # allauth will not work for the user later on.
-    if 'allauth.socialaccount' in settings.INSTALLED_APPS:
-        from allauth.socialaccount.models import SocialAccount, EmailAddress
-
-        if oidc:
-            provider_name = 'tampere_oidc'
-        else:
-            provider_name = 'tampere'
-        args = {'provider': provider_name, 'uid': user_id}
-        try:
-            account = SocialAccount.objects.get(**args)
-            assert account.user_id == user.id
-        except SocialAccount.DoesNotExist:
-            account = SocialAccount(**args)
-            account.extra_data = payload
-            account.user = user
-            account.save()
-
-            try:
-                email = EmailAddress.objects.get(email__iexact=user.email)
-                assert email.user == user
-            except EmailAddress.DoesNotExist:
-                email = EmailAddress(email=user.email.lower(), primary=True,
-                                     user=user, verified=True)
-                email.save()
+    try:
+        user = user_model.objects.select_for_update().get(uuid=user_id)
+    except user_model.DoesNotExist:
+        user = user_model(uuid=user_id)
+        user.set_unusable_password()
+
+    update_user(user, payload, oidc)
+    create_social_account(user, user_id, payload, oidc)
 
     return user
+
+
+def create_social_account(user, user_id, payload, oidc):
+    """Creates SocialAccount instance for this user if one
+    does not exist already.
+
+    Raises AuthenticationFailed if existing SocialAccount matching provider
+    and UUID for another user.
+
+    If user does not have an existing SocialAccount, will also attempt to create or match an
+    EmailAddress instance.
+    """
+    if not use_social_account:
```

Comment:
`use_social_account` will be undefined if `allauth.socialaccount` is not in `INSTALLED_APPS`.

Reply:
Should use_social_account be initialized as `false` before line 8?